In [2]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.svm import SVR
from sklearn.metrics import mean_squared_error, r2_score

In [3]:
data = '../Preprocessing/data.csv'

In [5]:
df = pd.read_csv(data)
df

Unnamed: 0  냉해노출수   총 강우량  1~2월 평균 누적 강수량     일조량  일사량     농가 수     재배면적  \
0        2019     20  1855.0           88.50  621560  982  13198.0  27689.0   
1        2020     11  2855.5          319.50  510738  801  10376.0  25372.0   
2        2021     21  2067.5           78.50  481655  708  11495.0  21999.0   
3        2022     26  1711.0            0.00  552621  810  11228.0  22362.0   
4        2023     24  3101.0           71.75  547410  816  10691.0  24700.0   
5        2024     16  1897.5          249.50  347284  466      NaN  23291.0   

   10a당 생산량       생산량  의성군재배면적   의성군생산량  
0    1400.0  387671.0   1700.0  19262.0  
1    1432.0  363432.0   1444.0  14937.0  
2    1402.0  312995.0   1340.0  13568.0  
3    1220.0  290824.0   1246.0  12084.0  
4    1288.0  312438.0   1215.0  11801.0  
5       NaN       NaN   1027.7      NaN

In [6]:
df.fillna(df.mean(), inplace=True)
df

Unnamed: 0  냉해노출수   총 강우량  1~2월 평균 누적 강수량     일조량  일사량     농가 수     재배면적  \
0        2019     20  1855.0           88.50  621560  982  13198.0  27689.0   
1        2020     11  2855.5          319.50  510738  801  10376.0  25372.0   
2        2021     21  2067.5           78.50  481655  708  11495.0  21999.0   
3        2022     26  1711.0            0.00  552621  810  11228.0  22362.0   
4        2023     24  3101.0           71.75  547410  816  10691.0  24700.0   
5        2024     16  1897.5          249.50  347284  466  11397.6  23291.0   

   10a당 생산량       생산량  의성군재배면적   의성군생산량  
0    1400.0  387671.0   1700.0  19262.0  
1    1432.0  363432.0   1444.0  14937.0  
2    1402.0  312995.0   1340.0  13568.0  
3    1220.0  290824.0   1246.0  12084.0  
4    1288.0  312438.0   1215.0  11801.0  
5    1348.4  333472.0   1027.7  14330.4

NaN값 평균값 대체

의성군 생산량 평균값 : 14330.4

In [24]:
df.iloc[-1,-1]=0
df

Unnamed: 0  냉해노출수   총 강우량  1~2월 평균 누적 강수량     일조량  일사량     농가 수     재배면적  \
0        2019     20  1855.0           88.50  621560  982  13198.0  27689.0   
1        2020     11  2855.5          319.50  510738  801  10376.0  25372.0   
2        2021     21  2067.5           78.50  481655  708  11495.0  21999.0   
3        2022     26  1711.0            0.00  552621  810  11228.0  22362.0   
4        2023     24  3101.0           71.75  547410  816  10691.0  24700.0   
5        2024     16  1897.5          249.50  347284  466  11397.6  23291.0   

   10a당 생산량       생산량  의성군재배면적   의성군생산량  
0    1400.0  387671.0   1700.0  19262.0  
1    1432.0  363432.0   1444.0  14937.0  
2    1402.0  312995.0   1340.0  13568.0  
3    1220.0  290824.0   1246.0  12084.0  
4    1288.0  312438.0   1215.0  11801.0  
5    1348.4  333472.0   1027.7      0.0

In [25]:
feature = df.iloc[:-1, 1:-1]
feature

냉해노출수   총 강우량  1~2월 평균 누적 강수량     일조량  일사량     농가 수     재배면적  10a당 생산량  \
0     20  1855.0           88.50  621560  982  13198.0  27689.0    1400.0   
1     11  2855.5          319.50  510738  801  10376.0  25372.0    1432.0   
2     21  2067.5           78.50  481655  708  11495.0  21999.0    1402.0   
3     26  1711.0            0.00  552621  810  11228.0  22362.0    1220.0   
4     24  3101.0           71.75  547410  816  10691.0  24700.0    1288.0   

        생산량  의성군재배면적  
0  387671.0   1700.0  
1  363432.0   1444.0  
2  312995.0   1340.0  
3  290824.0   1246.0  
4  312438.0   1215.0

In [26]:
target = df.iloc[:-1, -1]
target

0    19262.0
1    14937.0
2    13568.0
3    12084.0
4    11801.0
Name: 의성군생산량, dtype: float64

In [27]:
X_predict = df.iloc[-1, 1:-1].values.reshape(1, -1)
X_predict

array([[1.60000e+01, 1.89750e+03, 2.49500e+02, 3.47284e+05, 4.66000e+02,
        1.13976e+04, 2.32910e+04, 1.34840e+03, 3.33472e+05, 1.02770e+03]])

In [28]:
X_train, X_test, y_train, y_test = train_test_split(feature, target, test_size=0.2, random_state=42)


In [29]:
models = {
    "Linear Regression": LinearRegression(),
    "Decision Tree": DecisionTreeRegressor(),
    "Random Forest": RandomForestRegressor(),
    "Gradient Boosting": GradientBoostingRegressor(),
    "SVR": SVR()
}

In [32]:
# 모델 학습 및 예측
predictions = {}
for name, model in models.items():
    model.fit(feature, target)
    y_pred = model.predict(X_predict)
    predictions[name] = y_pred[0]
    print(f"{name}: {y_pred[0]}")

Linear Regression: 13764.274063217681
Decision Tree: 13568.0
Random Forest: 13815.95
Gradient Boosting: 14119.11689669293
SVR: 13568.003266610032


/Users/sml/anaconda3/envs/Torch_NLP38_/lib/python3.8/site-packages/sklearn/base.py:464: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(
/Users/sml/anaconda3/envs/Torch_NLP38_/lib/python3.8/site-packages/sklearn/base.py:464: UserWarning: X does not have valid feature names, but DecisionTreeRegressor was fitted with feature names
  warnings.warn(
/Users/sml/anaconda3/envs/Torch_NLP38_/lib/python3.8/site-packages/sklearn/base.py:464: UserWarning: X does not have valid feature names, but RandomForestRegressor was fitted with feature names
  warnings.warn(
/Users/sml/anaconda3/envs/Torch_NLP38_/lib/python3.8/site-packages/sklearn/base.py:464: UserWarning: X does not have valid feature names, but GradientBoostingRegressor was fitted with feature names
  warnings.warn(
/Users/sml/anaconda3/envs/Torch_NLP38_/lib/python3.8/site-packages/sklearn/base.py:464: UserWarning: X does not have valid feature names, but SVR was fitted 

In [33]:
predictions

{'Linear Regression': 13764.274063217681,
 'Decision Tree': 13568.0,
 'Random Forest': 13815.95,
 'Gradient Boosting': 14119.11689669293,
 'SVR': 13568.003266610032}

# 1차 회귀 분석 결과

- 'Linear Regression': 13764.27
- 'Decision Tree': 13568.0
- 'Random Forest': 13815.95
- 'Gradient Boosting': 14119.11
- 'SVR': 13568.00

## 맥에서 autoML이 안됨...ㅠ